### DISCUSSION TITLE
[SOLVED] How hard would it be to implement something like a custom RSS retriever for Twitter?

### DISCUSSION THREAD

**Discussion by felipe19930** [martinrotter/rssguard#354] [General]
I didn't want to open a new issue for this because it's mostly a question, but I wonder if RSS Guard provides an easy interface to create custom feeds for websites that don't provide a RSS interface at all, like Twitter.

Thanks!

**Comment by martinrotter** [comment]
Hello @felipe19930. Discussions are exactly for these kinds of talks, okay.

Generally, at this point we have several options and if they are feasible depends on how exactly you would want the feature to behave:
1. I could implement https://github.com/martinrotter/rssguard/issues/265 which would allow you to do this:
    1. Download some fixed URL contents either directly automatically via RSS Guard or via calling external arbitrary script (with parameter).
    2. Process downloaded data with custom script inside RSS Guard - the script could "convert" data to ATOM/RSS XML format.
    3. Process resulting XML with standard "plugin".
2. Or we could write custom service plugin specifically designed to work with Twitter, probably via [Twitter API](https://developer.twitter.com/en/docs).

Difference is that 1. is more general and very flexible solution which will allow you to process all kinds of websites w/o feeds, convert them to feeds and then process feeds while option 2. is specifically designed to only work with Twitter (in a similar fashion like other already existing plugins like Inoreader or Nextcloud News plugin).

Option 1. is meant to be programmed by me and then configured by users, option 2. must be entirely programmed either by me or by some contributor - at least basic knowledge of C++/Qt is needed.

Note that if you would require some advanced Twitter-related functionality like adding +1 to posts or maybe commenting (?) on posts, only option 2. is viable. Also note that there is already existing well-defined RSS Guard plugin API for writing new "plugins" that every at least semi-competent C++ programmer  could use.

**Reply by martinrotter** [reply]
Follow up: Option 1. will allow to use Tweeper which is special RSS feed generator for Twitter. Resulting XML file produced by Tweeper will of course work with other RSS readers probably too, if properly configured or hacked.

**Reply by martinrotter** [reply]
Excerpt of Tweeper's readme:
```
  41 Tweeper can be used via web or as a command line program, for example as
  42 a filter in your feed reader, by passing the URL of the user's public timeline
  43 as the first argument.
  44 
  45 Example of use on the command line:
  46 
  47   $ php tweeper.php https://twitter.com/NSACareers
```

**Reply by martinrotter** [reply]
https://git.ao2.it/tweeper.git/blob/master:/README

**Reply by martinrotter** [reply]
Btw, in the meantime, I started to work on solution 1 - see #265 to track progress. I will add some examples with twitter and document the whole thing properly later in the documentation.

**Reply by felipe19930** [reply]
Awesome! Thanks for such a in-depth response too.

>Note that if you would require some advanced Twitter-related functionality like adding +1 to posts or maybe commenting (?) on posts, only option 2. is viable. 

My use-case is very simple actually: There are some Twitter accounts I want to "follow" (i.e. monitor their tweets), because they often post shopping coupons (or great deals in general), so I'm interested in getting those as quick as possibl,e before they expire. Right now, I'm using a [nitter instance](https://github.com/zedeus/nitter/wiki/Instances) to get these tweets in the RSS format, and while it's been working great for a couple of months now, nitter appears to have a 10 minute delay or something before a new tweet appears, which is not _that_ interesting for my use-case.

I like option 1 because, like you said, would also work for other sites, and I know a few that this feature would come in handy.

Tweeper sounds like a great alternative once #265 gets implemented, so I'll start getting to know it better!

Thanks for starting the implementation too, because if we were to depend on my C++ knowledge to do anything, then nothing would be done. 😅

**Reply by martinrotter** [reply]
@felipe19930 Yes, I am implementing the ticket #265 right now and it is quite serious stuff. I will commit some code soon and expect it to be done in a week. I have actually tried Tweeper now as part if my internal testing and I must say that it works nicely. It takes URL of feed as input and produces RSS feed XML file, although it was quite tricky to install it on Windows. I had to install XAMPP (to get "php.exe"), then I had to install Composer, then install Symfony via Composer, then install Tweeper.

Could you move your next messages directly to #265 as we in fact "solved" this discussion. :)

**Comment by martinrotter** [comment]
@felipe19930 OK, raw functionality is there. You can now download latest development build and test the feature.

https://github.com/martinrotter/rssguard/blob/master/resources/docs/Feed-formats.md#websites-scraping-and-related-advanced-features

Note that the feature was so far tested only under Windows. I will test with Linux on the evening within a week. If you have any questions, ask.

https://github.com/martinrotter/rssguard/issues/265

**Reply by felipe19930** [reply]
Wow, that was quick! Thanks :)

**EDIT:** Just to correct my post below, although the devbuild dates back to December 2020, the build itself is from the latest commit. However, no AppImage is available there :(

I definitely want to give it a try, but it seems like the latest devbuild available is very outdated. And the GitHub CI doesn't seem to provide build artifacts either, unless I completely missed them... I could build it myself but my laptop isn't that powerful so if it takes 10 minutes for the Ubuntu build to be compiled on GitHub servers, it will take an hour or more on my laptop :D

**Reply by martinrotter** [reply]
Yes, there are development builds which are compiled automatically after EACH commit, they are super new.

https://github.com/martinrotter/rssguard/releases/tag/devbuild

Do not mind that it says "released this on Dec 8, 2020 · 165 commits to master since this release", always check the name and version of the file directly, for example "rssguard-3.8.4-f24f03e6-win64.7z". Note that the weird string "f24f03e6" indicates the "commit" of the build and you can search for the commit using the left/top search textbox here on github. Note that Linux builds are now broken, I will fix them.

f24f03e6

**Reply by felipe19930** [reply]
Just to give a feedback of this feature on Linux: 

I tried using the `curl` example you provided in the README, but it doesn't work. This is my Script source:

```
bash#-c "curl 'https://github.com/martinrotter.atom'"
```

![image](https://user-images.githubusercontent.com/35700659/106821188-72dbda00-665b-11eb-9d34-4be05db1e210.png)


After clicking "Fetch it now", I get the following message in the terminal:

```
-> core: Running custom script for guessing 'bash#-c "curl 'https://github.com/martinrotter.atom'"' to obtain feed data.
-> core: XML of feed 'bash#-c "curl 'https://github.com/martinrotter.atom'"' is not valid and cannot be loaded. Error: 'unexpected end of file' (line 1, column 1).
```

However, if I change the script source to this instead:

```
curl#https://github.com/martinrotter.atom
```

Then it works. But whenever I try to pass more arguments to my command, it stops working again. For example:

```
curl#-s https://github.com/martinrotter.atom
```

Then I get this:

```
-> core: Running custom script for guessing 'curl#-s https://github.com/martinrotter.atom' to obtain feed data
-> core: XML of feed 'curl#-s https://github.com/martinrotter.atom' is not valid and cannot be loaded. Error: 'unexpected end of file' (line 1, column 1).
```

I'm not an expert, but after using `strace` to monitor all `execve()` calls RSS Guard makes, it appears that there's something going wrong with how the command is executed. To test this, first I had to extract the AppImage, because `strace` wasn't working when I tried to run it directly:

```
$ /rssguard-3.8.4-a9efaac8-nowebengine-linux64.AppImage --appimage-extract
```

Then I ran RSS Guard:

```
$ QT_LOGGING_RULES='*rssguard*=true' strace -f -e execve ./squashfs-root/AppRun
```

Then I tried using that first `curl` example I mentioned in the beginning again, and got this information after clicking "Fetch it now" (timestamps were removed for readabilty):

```
-> core: Running custom script for guessing 'bash#-c "curl 'https://github.com/martinrotter.atom'"' to obtain feed data.
strace: Process 51220 attached
[pid 51220] execve("/usr/bin/bash", ["/usr/bin/bash", "-c \"curl 'https://github.com/mar"...], 0xf9f8a0 /* 72 vars */) = 0
[pid 51220] +++ exited with 2 +++
[pid 51157] --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=51220, si_uid=1000, si_status=2, si_utime=0, si_stime=0} ---
-> core: XML of feed 'bash#-c "curl 'https://github.com/martinrotter.atom'"' is not valid and cannot be loaded. Error: 'unexpected end of file' (line 1, column 1).
```

Now if we run the following command directly from the terminal:

```
$ strace -f -e execve bash -c "curl 'https://github.com/martinrotter.atom'" 1>/dev/null
execve("/usr/bin/bash", ["bash", "-c", "curl 'https://github.com/martinr"...], 0x7ffed247ec58 /* 69 vars */) = 0
execve("/usr/bin/curl", ["curl", "https://github.com/martinrotter."...], 0x55b7a7ab7170 /* 70 vars */) = 0
strace: Process 51643 attached
[pid 51643] +++ exited with 0 +++
+++ exited with 0 +++
```

We can immediately notice how the arguments are properly split in the first line when compared to RSS Guard's output.

I still have no idea how to fix it, though, but that's a start, I guess. :D

**Reply by martinrotter** [reply]
@felipe19930 Wow, thanks for testing. Some amazing output you got. Anyway, as I said I did not even test it on Linux yet. Will do some extensive testing tonight. I know the source of the bug you receive. There is one HUGE GIANT flaw in C++ libraries when it comes to process launching.

For some crazy reason, they DO NOT allow me to pass a single "argument" line/string. The API accepts only LIST OF STRINGS and they actually expect you to "split" arguments into the list. For example argument line `-c "curl aaaabbbb.com/file"` is expected to be called via API like this: `blablbla `-c` `"curl aaaabbbb.com/file"`. Fuck me, right? I will test things out and maybe maybe I will be forced to use Linux native IPC for this.

https://doc.qt.io/qt-5/qprocess.html#setArguments
https://doc.qt.io/qt-5/qprocess.html#setNativeArguments (Windows only, WTF).

**Reply by felipe19930** [reply]
LOL. Dear God, that sounds like a nightmare :(

Hope you can find a way though

**Reply by martinrotter** [reply]
I will. There are some cases I like to avoid. One of them is using non cross-platform APIs. I would like to strictly stick to Qt APIs. I see possible solution to actually enhance "execution line" format and use `#` separator to actually separate all arguments. Because the real problem here is that it is very hard to split single argument string into separate arguments because sometimes it is not that clear what is meant to be single argument (for example quoted string with spaces) and what is not.

My suggestion is this change, instead of `bash#-c "curl 'https://github.com/martinrotter.atom'"` to have `bash#-c#curl 'https://github.com/martinrotter.atom'`, which could with some minor internal changes work. I will test tonight.

**Reply by martinrotter** [reply]
OK, right now, I did very basic test and made some changes and tested this execution line:

```
C:\Programy\xampp\php\php.exe#C:\Projekty\Moje\tweeper\tweeper.php#-v#0#https://twitter.com/NSACareers
```
Note that `#` is now used to delimit each single argument. That way you can say exactly which arguments are separate. Also, there is almost no need to quote arguments like strings, but this depends quite a lot on the exact utility used as a script interpreter.

**Reply by martinrotter** [reply]
@felipe19930 OK I pushed the changes and development build should arrive there in a 15 minutes. I updated documentation with newest examples which worked for me:

https://github.com/martinrotter/rssguard/blob/master/resources/docs/Feed-formats.md#websites-scraping-and-related-advanced-features

and tested on Windows. I believe that current implementation could also work on Linux and Mac OS but will test tonight.

**Reply by felipe19930** [reply]
That seems a little convoluted, in my humble opinion. Also, what if for some reason I needed to pass the literal `#` character as an argument? Would it work?

Anyway, I've looked up on how QuiteRSS does it (well, QuiteRSS doesn't have this feature). To be more specific, I've searched on how their "Open in external browser" feature works, and it appears to handle extra arguments just fine. I have this on my QuiteRSS' settings:

![Screenshot from 2021-02-04 05-52-27](https://user-images.githubusercontent.com/35700659/106868182-3afd8280-66ad-11eb-9574-aff17ddeb49e.png)

And using `strace` to monitor it, QuiteRSS appears to handle every argument pretty well:

![image](https://user-images.githubusercontent.com/35700659/106868089-1c978700-66ad-11eb-8442-25edeb39f123.png)

And I assume this is the code that handles the "Open in external browser" action:

https://github.com/QuiteRSS/quiterss/blob/271c55756dcf19ca163a603a54d12f165548a602/src/newstabwidget.cpp#L2198

Since both projects share the same license, perhaps you could borrow QuiteRSS's code (please forgive me if I'm talking non-sense and this is not applicable for RSS Guard 😅)

**Reply by martinrotter** [reply]
EDIT: I checked their code. And it is non-usable for our use-case because in their use-case, they simply launch external command and do not care about it, but in our use case we need to grab command's standard output and also in same cases provided data to command's standard input. Moreover, they use windows-specific APIs on Windows.

 As for now, I tested the last code on Linux:
![image](https://user-images.githubusercontent.com/1255302/106870644-a0636a80-66d1-11eb-9d39-a236fd05148e.png)

It works, but I will check their code.

**Reply by felipe19930** [reply]
It's okay, if it works, it works!

Thanks a lot ❤️

**Reply by martinrotter** [reply]
@felipe19930 As for "#" character. Yes, sadly, there is probably no way of using it in your command string. You could probably use it if encoded, for example like this in Bash: `echo -e '\u0023'` -> "#".

We can certainly pick a different character as separator, just suggest one.

**Comment by CsatiZoltan** [comment]
What about creating an interface to RSS feed generators that can create a feed link to almost any webpage? Something like https://rss.app/ or a free alternative.

**Reply by martinrotter** [reply]
@CsatiZoltan At this point, features related to "scraping" are implemented and I tested them like hell. They work nicely.

As for Twitter, I believe there is like 10+ ways of getting Twitter RSS/ATOM/JSON feeds. You can use

https://git.ao2.it/tweeper.git/blob/master:/README

which is also written in [docs](https://github.com/martinrotter/rssguard/blob/master/resources/docs/Feed-formats.md#websites-scraping-and-related-advanced-features), specifically this line `php#tweeper.php#-v#0#https://twitter.com/NSACareers` in RSS Guard feed's settings dialog in "source" attribute with "script" option will do everything needed.

Also, you can probably just scrape the twitter with generic Python-based script with some regexp magic in a similar fashin like [these](https://github.com/martinrotter/rssguard/tree/version-4/resources/scripts/scrapers).

As for rss.app if I understand it correctly, it should already work with RSS Guard and pretty much every other feed reader at this point.

> creating an interface to RSS feed generators

Could you specify exactly what and how should this interface look/work?

**Reply by CsatiZoltan** [reply]
@martinrotter 



> Could you specify exactly what and how should this interface look/work?

I don't have a firm idea, I was just thinking that you could add a push button in the "Add new feed" window that opens a dialogue box, where the user must give the link for which he/she wants the feed. Then RSS Guard would call *rss.app* or another scraper and fetch the link they generate. So it would be a convenience option that the user wouldn't have to do that manually on *rss.app* and paste the link, but RSS Guard would do that in the background. Similarly to what it does with *tweeper* (see the issue below).

For curiosity, I tried the line `php#tweeper.php#-v#0#https://twitter.com/NSACareers`. It gave me the following error:
```
time="   185.768" type="debug" -> core: Running custom script for guessing 'php#tweeper.php#-v#0#https://twitter.com/NSACareers' to obtain feed data.
time="   185.806" type="debug" -> core: XML of feed 'php#tweeper.php#-v#0#https://twitter.com/NSACareers' is not valid and cannot be loaded. Error: 'error occurred while parsing element' (line 1, column 1).
```

> As for rss.app if I understand it correctly, it should already work with RSS Guard and pretty much every other feed reader at this point.

Yes, it works, but it only supports [two links for a free plan](https://rss.app/plans).

**Reply by martinrotter** [reply]
@CsatiZoltan In order for `php#tweeper.php#-v#0#https://twitter.com/NSACareers` to work, you need to have `PHP` installed and configured, as well as Tweeper itself.

![image](https://user-images.githubusercontent.com/1255302/111422534-24f0d200-86ef-11eb-98a1-a560dfefaf70.png)

The error message you got was bit misleading, fixed that little nuance. Should report better error messages now.

**Reply by martinrotter** [reply]
@CsatiZoltan  As for the idea itself, I believe that the intended feature is out-of-scope for RSS Guard. :) I do not want to add yet another weak dependency to maintain. :)

**Reply by CsatiZoltan** [reply]
> @CsatiZoltan As for the idea itself, I believe that the intended feature is out-of-scope for RSS Guard. :) I do not want to add yet another weak dependency to maintain. :)

Completely understand, I also strive for minimal dependencies in my projects...
